### PR TITLE
Avoid using DialogPage.Save/LoadSettings

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
For 0.5 changes are not complete. I'd like to clean up names, add tests to the settings persistence as well as remove `Browsable(false)` attributes. This is to be done in master. For Dev15 we only want to avoid using dialog page for storage/loading. Codee may still call SaveSettings on the dialog page, but implementation is via VS settings manager.

#2618

### CODE REVIEW

File: src/Package/Impl/Options/R/RToolsOptionsPage.cs [hunk 1]
```diff
@@ -291,18 +294,37 @@ public class RToolsOptionsPage : DialogPage {
...
             base.OnApply(e);
         }
+
+        private void PersistSettings() {
+            SaveSettingsToStorage();
+            _storage.Persist();
+        }
+
+        protected override void LoadSettingFromStorage(PropertyDescriptor prop) {
+            if (_storage.SettingExists(prop.Name)) {
```
Comment (AlexanderSher) [inline src/Package/Impl/Options/R/RToolsOptionsPage.cs]: So we intentionally don't call base method here?

File: src/Package/Impl/Shell/VsSettingsStorage.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    [Export(typeof(ISettingsStorage))]
+    internal sealed class VsSettingsStorage : ISettingsStorage {
+        private const string _collectionPath = "R Tools";
+        private readonly Dictionary<string, object> _settings = new Dictionary<string, object>();
+        private readonly ShellSettingsManager _shellSettingsManager;
+        private readonly WritableSettingsStore _store;
+        private readonly object _lock = new object();
+
+        public VsSettingsStorage() {
+            _shellSettingsManager = new ShellSettingsManager(RPackage.Current);
```
Comment (AlexanderSher) [inline src/Package/Impl/Shell/VsSettingsStorage.cs]: `ShellSettingsManager` - is it visual studio type?

File: src/Package/Impl/Options/R/RToolsOptionsPage.cs (hunk 1 above)
Comment (MikhailArkhipov) [inline src/Package/Impl/Options/R/RToolsOptionsPage.cs]: Yes. I don't want to mess with this code too in 0.5 - it already handles property names. What really should be done is to remove save/load calls from here, have own table of property names for each setting (or use approach similar to `BindableBase.SetProperty` with caller name attribute) and only go via setting storage.  So `RPackage.LoadSettings` won't be calling `RToolsOptionsPage.LoadSettings`

File: src/Package/Impl/Shell/VsSettingsStorage.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            lock (_lock) {
+                _settings[name] = value;
+            }
+        }
+
+        public void Persist() {
+            lock (_lock) {
+                foreach (var s in _settings) {
+                    var t = s.Value.GetType();
+                    if (typeof(bool).IsAssignableFrom(t)) {
```
Comment (AlexanderSher) [inline src/Package/Impl/Shell/VsSettingsStorage.cs]: `s.Value is bool`?

File: src/Package/Impl/Shell/VsSettingsStorage.cs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            Debug.Assert(_store != null);
+
+            if (!_store.CollectionExists(_collectionPath)) {
+                _store.CreateCollection(_collectionPath);
+            }
+        }
+
+        public bool SettingExists(string name) {
+            lock (_lock) {
+                if (_settings.ContainsKey(name)) {
```
Comment (AlexanderSher) [inline src/Package/Impl/Shell/VsSettingsStorage.cs]: Maybe just add a comment?

File: src/Package/Impl/Shell/VsSettingsStorage.cs (hunk 2 above)
Comment (AlexanderSher) [inline src/Package/Impl/Shell/VsSettingsStorage.cs]: Ok

File: src/Package/Impl/Shell/VsSettingsStorage.cs (hunk 4 above)
Comment (MikhailArkhipov) [inline src/Package/Impl/Shell/VsSettingsStorage.cs]: sure

File: src/Package/Impl/Shell/VsSettingsStorage.cs [hunk 5]
```diff
@@ -67,21 +85,21 @@ internal sealed class VsSettingsStorage : ISettingsStorage {
...
 
         public void Persist() {
             lock (_lock) {
-                foreach (var s in _settings) {
+                foreach (var s in _settingsCache) {
                     var t = s.Value.GetType();
-                    if (typeof(bool).IsAssignableFrom(t)) {
+                    if (s.Value is bool) {
                         _store.SetBoolean(_collectionPath, s.Key, (bool)s.Value);
                     } else if (typeof(int).IsAssignableFrom(t) || t.IsEnum) {
```
Comment (AlexanderSher) [inline src/Package/Impl/Shell/VsSettingsStorage.cs]: `s.Value is int || s.Value is Enum`

**Comment by AlexanderSher** [comment]
👍

File: src/Package/Impl/Shell/VsSettingsStorage.cs (hunk 5 above)
Comment (AlexanderSher) [inline src/Package/Impl/Shell/VsSettingsStorage.cs]: `typeof(int).IsAssignableFrom(typeof(uint))` and `typeof(uint).IsAssignableFrom(typeof(int))` both return `false`.

Comment (MikhailArkhipov) [inline src/Package/Impl/Shell/VsSettingsStorage.cs]: I guess since compiler doesnt like such assignments either. :-) Will add separate clause in master.

**Pull request merged**